src/cljs/main/broadfcui/common/table/table.cljs

```diff
@@ -0,0 +1,141 @@
+(ns broadfcui.common.table.table
+  (:require
+    [clojure.set :refer [difference subset?]]
+    [dmohs.react :as react]
+    [broadfcui.common.components :as comps]
+    [broadfcui.common.table.body :as body]
+    [broadfcui.common.table.column-editor :refer [ColumnEditButton]]
+    [broadfcui.common.table.paginator :as paginator]
+    [broadfcui.common.table.utils :as table-utils]
+    [broadfcui.common.style :as style]
+    [broadfcui.persistence :as persistence]
+    [broadfcui.utils :as utils]
+    ))
+
+
+;; Define default props this way because we need to do a deep-merge,
+;; instead of React's regular merge.
+(def ^:private default-props
+  {:body {:empty-message "There are no rows to display."
+          :external-query-params #{}
+          :behavior {:reorderable-columns? true
+                     :fixed-column-count 0
+                     :sortable-columns? true
+                     :allow-no-sort? false
+                     :resizable-columns? true
+                     :filterable? true}}
+   :toolbar {:style {:display "flex" :alignItems "baseline" :marginBottom "1rem"}
+             :column-edit-button {:style {:marginRight "1rem"}
+                                  :button {:text "Columns" :icon :settings
+                                           :style {:padding "0.4rem 0.8rem 0.4rem 0.4rem"}}
+                                  :anchor :left}
+             :filter-bar {:style {:marginRight "1rem"}}}
+   :paginator {:style {:marginTop "1rem"}
+               :per-page-options [10 20 100 500]}})
+
+(def ^:private all-query-params #{:page-number :rows-per-page :filter-text :sort-column :sort-order})
+
+
+(react/defc Table
+  {:update-query-params
+   (fn [{:keys [state]} new-params]
+     (assert (subset? (set (keys new-params)) all-query-params) "Unknown key passed to :update-query-params")
+     (let [old-state (:query-params @state)
+           new-state (merge old-state new-params)
+           different? (not= old-state new-state)]
+       (when different?
+         (swap! state assoc :query-params new-state))
+       different?))
+   :refresh-rows
+   (fn [{:keys [props state]}]
+     (swap! state assoc :loading? true)
+     ((-> props :body :data-source) {:columns (-> props :body :columns)
```

Comment:
`:data-source` is a function that gets the query params and returns the appropriate rows, either via ajax or via client-side manipulation.